### DISCUSSION TITLE
Add `retrieveCookie` to the `AsyncPlayerPreLoginEvent` (or any other way to query cookies during the login phase)

### DISCUSSION THREAD

**Discussion by MrPowerGamerBR** [PaperMC/Paper#11734] [Ideas]
### Is your feature request related to a problem?

Currently you can't get cookies without a player instance, even tho you are able to get cookies during the login phase.

If you add a `this.sendPacket(new ClientboundCookieRequestPacket(ResourceLocation.parse("...")));` to the `callPlayerPreLoginEvents` and check if the payload is present on the `handleCookieResponse`, you can see that the client does send the cookie response and it does send it with the payload.

### Describe the solution you'd like.

Add `retrieveCookie` to the `AsyncPlayerPreLoginEvent`. (or another API that allows to query the player's cookies)

### Describe alternatives you've considered.

Using a proxy /j

**Comment by electronicboy** [comment]
Events are just POJOs; to do this, you'd need to block the event, which would, in turn, stall up a thread. This needs new API to properly handle this phasing, and ideally a proper system for this akin to what mojang does with their tasks for configuration, rather than just making people block arbitrary events, if we wanted to support that, we'd need to support continuations or something

**Reply by KaiKikuchi** [reply]
At the moment, if we want to get a cookie during the login process, we have to stall the PlayerLoginEvent, which is way worse than stalling the AsyncPlayerPreLoginEvent. There is a reason we want to access cookies data during the login process: we might want to decide whether to reject a player based on a cookie, and it would be much better if we could do this before player authentication. This is also particularly useful if we could, for instance, skip the player's authentication process and just let them in onto the server, so that every time we transfer a player between our servers, we won't have to make them wait for authentication.

I am proposing a simple solution to this: plugins register one or more namespaced key identifying one or more cookie to the Bukkit/Paper API (maybe similar to how we do it with plugin channels, i.e. `Bukkit.getCookieManager().registerPreLoginCookie(NamespacedKey key)`, with key being the key identifying the cookie) so that just before the AsyncPlayerPreLogin event those cookies are being retrieved from the client, making them available on the AsyncPlayerPreLogin event with a "getRegisteredCookie(NamespacedKey key)" method, returning the cookie data, or null if the client didn't have the cookie set.

**Reply by Owen1212055** [reply]
This will be resolved by the confirmation stage api.

**Reply by Owen1212055** [reply]
For reference— this exposes a player connection object to certain events that allow you to run methods like the fetch cookie.

**Reply by KaiKikuchi** [reply]
Thanks for the quick reply. That's great news. I hope I can use that soon, it would solve all my issues!
Is there any chance we can get a method to decide whether to run the session authentication with Mojang on the AsyncPlayerPreLogin event? That way, we can allow direct connections to the server by properly authenticating players, and at the same time have players being transferred from other trusted servers not authenticate with Mojang (plugin devs will handle this properly with session cookies). Thanks!

EDIT: while we wait for proper API, I made a fork and added the "retrieveCookie(String key)" method to AsyncPlayerPreLoginEvent

**Reply by Owen1212055** [reply]
Most likely a separate event could be added which occurs even earlier in the login process.
